File: quiz-web/src/api/practice.ts
import { http, buildQuery } from "./client";

export type SessionMode = "practice" | "exam" | "mistake" | "category";
export type SessionStatus = "active" | "paused" | "submitted" | "abandoned";

export interface SessionFilter {
  category_ids?: number[];
  types?: string[];
  difficulties?: number[];
  tags?: string[];
}

export interface PracticeSessionCreate {
  mode?: SessionMode;
  title?: string | null;
  count?: number;
  duration_limit_sec?: number | null;
  filter?: SessionFilter;
}

export interface QuestionPlay {
  id: number;
  code: string;
  type: string;
  stem: string;
  difficulty: number;
  category_id?: number | null;
  category_name?: string | null;
  tags: string[];
  payload: Record<string, any>;
  status: string;
  source?: string | null;
  version: number;
  created_at: string;
  updated_at: string;
}

export interface SessionItem {
  id: number;
  session_id: number;
  question_id: number;
  seq: number;
  score_weight: number;
  flagged: boolean;
  answered: boolean;
  is_correct?: boolean | null;
  score?: number | null;
  spent_sec: number;
  question: QuestionPlay;
  reveal?: Record<string, any> | null;
}

export interface PracticeSession {
  id: number;
  code: string;
  title?: string | null;
  mode: SessionMode;
  status: SessionStatus;
  total_count: number;
  duration_limit_sec?: number | null;
  elapsed_sec: number;
  started_at?: string | null;
  last_active_at?: string | null;
  submitted_at?: string | null;
  score?: number | null;
  correct_count?: number | null;
  accuracy?: number | null;
  items: SessionItem[];
}

export interface PracticeSessionListItem {
  id: number;
  code: string;
  title?: string | null;
  mode: SessionMode;
  status: SessionStatus;
  total_count: number;
  correct_count?: number | null;
  accuracy?: number | null;
  started_at?: string | null;
  submitted_at?: string | null;
}

export interface AnswerSubmit {
  item_id: number;
  response: Record<string, any>;
  duration_sec?: number | null;
  self_eval?: "mastered" | "fuzzy" | "unknown" | null;
}

export interface AnswerBatchResult {
  item_id: number;
  ok: boolean;
  is_correct?: boolean | null;
  correct?: boolean;
  error?: string;
}

export interface AnswerResult {
  item_id: number;
  is_correct?: boolean | null;
  correct: boolean;
  score: number;
  max_score: number;
  judge_detail: Record<string, any>;
  need_manual: boolean;
  feedback?: string | null;
  reveal: Record<string, any>;
  progress: { answered: number; total: number; correct_count: number };
}

export interface SessionSubmitResult {
  id: number;
  status: SessionStatus;
  score?: number | null;
  correct_count: number;
  accuracy?: number | null;
  total_count: number;
  elapsed_sec: number;
}

export interface SessionListPage {
  items: PracticeSessionListItem[];
  total: number;
  page: number;
  page_size: number;
}

export const practiceApi = {
  create: (d: PracticeSessionCreate) => http.post<PracticeSession>("/practice/sessions", d),
  list: (q: { mode?: string; status?: string; page?: number; page_size?: number } = {}) =>
    http.get<SessionListPage>(`/practice/sessions${buildQuery(q as Record<string, unknown>)}`),
  get: (id: number) => http.get<PracticeSession>(`/practice/sessions/${id}`),
  patch: (id: number, d: { status?: SessionStatus; elapsed_sec?: number }) =>
    http.patch<PracticeSession>(`/practice/sessions/${id}`, d),
  answer: (id: number, d: AnswerSubmit) =>
    http.post<AnswerResult>(`/practice/sessions/${id}/answer`, d),
  answerBatch: (id: number, items: AnswerSubmit[]) =>
    http.post<{ results: AnswerBatchResult[] }>(`/practice/sessions/${id}/answer-batch`, {
      items,
    }),
  submit: (id: number) => http.post<SessionSubmitResult>(`/practice/sessions/${id}/submit`),
};

File: quiz-server/app/schemas/practice.py
"""练习会话（M3）相关入参/出参，见 DESIGN §5。

数据流：建会话（按筛选抽题）→ 逐题作答（即时判分、写 attempt/stat/mistake）
→ 交卷（汇总分数/正确率）。出参刻意把题干与选项（payload）和答案（answer）
分离：未作答前前端只拿到题目，作答后才在 reveal 里返回答案与解析。
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from app.schemas.common import ORMBase
from app.schemas.enums import QuestionType, SessionMode, SessionStatus


class SessionFilter(BaseModel):
    """抽题筛选条件。"""

    category_ids: list[int] = Field(default_factory=list)
    types: list[QuestionType] = Field(default_factory=list)
    difficulties: list[int] = Field(default_factory=list)
    tags: list[str] = Field(default_factory=list)


class PracticeSessionCreate(BaseModel):
    mode: SessionMode = SessionMode.practice
    title: str | None = Field(default=None, max_length=128)
    count: int = Field(default=20, ge=1, le=200)
    duration_limit_sec: int | None = Field(default=None, gt=0)
    filter: SessionFilter = Field(default_factory=SessionFilter)


class QuestionPlayOut(ORMBase):
    """题干展示用，刻意不含 answer / analysis / judge_config。"""

    id: int
    code: str
    type: QuestionType
    stem: str
    difficulty: int
    category_id: int | None = None
    category_name: str | None = None
    tags: list[str] = Field(default_factory=list)
    payload: dict[str, Any] = Field(default_factory=dict)
    status: str
    source: str | None = None
    version: int
    created_at: datetime
    updated_at: datetime


class SessionItemOut(ORMBase):
    id: int
    session_id: int
    question_id: int
    seq: int
    score_weight: float = 1.0
    flagged: bool = False
    answered: bool = False
    is_correct: bool | None = None
    score: float | None = None
    spent_sec: int = 0
    question: QuestionPlayOut
    # 作答后揭示：答案 + 解析 + 判分细节
    reveal: dict[str, Any] | None = None


class PracticeSessionOut(ORMBase):
    id: int
    code: str
    title: str | None = None
    mode: SessionMode
    status: SessionStatus
    total_count: int = 0
    duration_limit_sec: int | None = None
    elapsed_sec: int = 0
    started_at: datetime | None = None
    last_active_at: datetime | None = None
    submitted_at: datetime | None = None
    score: float | None = None
    correct_count: int | None = None
    accuracy: float | None = None
    items: list[SessionItemOut] = Field(default_factory=list)


class PracticeSessionListItem(ORMBase):
    id: int
    code: str
    title: str | None = None
    mode: SessionMode
    status: SessionStatus
    total_count: int = 0
    correct_count: int | None = None
    accuracy: float | None = None
    started_at: datetime | None = None
    submitted_at: datetime | None = None


class AnswerSubmit(BaseModel):
    """一次作答。客观题只需 response；简答需带 self_eval 自评。"""

    item_id: int
    response: dict[str, Any] = Field(default_factory=dict)
    duration_sec: int | None = Field(default=None, ge=0)
    self_eval: str | None = Field(default=None, pattern="^(mastered|fuzzy|unknown)$")


class AnswerBatch(BaseModel):
    """整卷批量作答：一组作答一次性提交（考试模式交卷用）。"""

    items: list[AnswerSubmit] = Field(default_factory=list)


class AnswerResult(BaseModel):
    item_id: int
    is_correct: bool | None = None
    correct: bool = False
    score: float = 0.0
    max_score: float = 1.0
    judge_detail: dict[str, Any] = Field(default_factory=dict)
    need_manual: bool = False
    feedback: str | None = None
    reveal: dict[str, Any] = Field(default_factory=dict)
    progress: dict[str, Any] = Field(default_factory=dict)


class SessionSubmitResult(BaseModel):
    id: int
    status: SessionStatus
    score: float | None = None
    correct_count: int = 0
    accuracy: float | None = None
    total_count: int = 0
    elapsed_sec: int = 0


class SessionPatch(BaseModel):
    """暂停/恢复、累加用时。"""

    status: SessionStatus | None = None
    elapsed_sec: int | None = Field(default=None, ge=0)
